### ISSUE TITLE
Sequential ransac detections for multiplane extraction

### ISSUE THREAD

**Issue by miguelriemoliveira** [miguelriemoliveira/RansacPlaneDetection#2]
Tiago can handle this I think ...

**Comment by tiagomfmadeira** [comment]
![multiple_plane_detection](https://user-images.githubusercontent.com/15963210/56921246-30b45a80-6abd-11e9-8646-196e3f9d19ea.png)

**Comment by tiagomfmadeira** [comment]
In this version I used Open3d to load the point cloud files and take advantage of their visualisation features.

**Comment by tiagomfmadeira** [comment]
Except sometimes, something like this happens...
![Uploading Screenshot from 2019-04-29 20-33-59.png…]()

**Comment by tiagomfmadeira** [comment]
Except sometimes, something like this happens...
![Screenshot from 2019-04-29 20-33-59](https://user-images.githubusercontent.com/15963210/56921743-7b82a200-6abe-11e9-9776-f525743a98b4.png)

**Comment by tiagomfmadeira** [comment]
The particular case with blue is similar to what you showed me the other day @miguelriemoliveira and could maybe be solved by clustering... but the green and red planes we don't want at all...

**Comment by miguelriemoliveira** [comment]
Hi Tiago,

red and green we can discard using a parameter for the minimum number of inliers.

The blue, I suggest the following procedure:
After detecting the inliers, before refining the plane coefficients, do a clustering procedure that separates the inliers into N groups, and keep only as inliers the group which contains more points.

Clustering based on point to point distance is often called Euclidean distance clustering, here is an [example in c++](http://www.pointclouds.org/documentation/tutorials/cluster_extraction.php).

Our goal is to have this is python, you can check python-pcl, which did not work for plane segmentation but I think was working for other examples, perhaps the clustering works and you can use it off the shelve.

https://github.com/strawlab/python-pcl/blob/master/examples/official/Segmentation/cluster_extraction.py

**Comment by miguelriemoliveira** [comment]
> red and green we can discard using a parameter for the minimum number of inliers.

by the way, which parameters you have now?

number_iterations, distance_threshold, minumum_number_of_points ...

any others?

**Comment by tiagomfmadeira** [comment]
I'm only using:
number_iterations = 100
number_points = points.shape[0]
distance_threshold = 0.08
(check [this](https://github.com/miguelriemoliveira/RansacPlaneDetection/blob/359674aaf5ee870ccbeb161620d64a8fb274c929/test/test_ransac_multiplane.py#L106) out)

I guess I could try to check for minimum number of inliners.

As far as clustering based on point to point distance goes, I remember from our conversation the other day. The example you showed me was somewhat similar with the fences(?) on the street. 
I'll have to look into it. Unfortunately, I think using python-pcl is gonna be a problem... because I can't seem to be able to install it. Not using pip nor from source...

**Comment by tiagomfmadeira** [comment]
However, let's suppose I do discard red and green using a parameter for the minimum number of inliers. What then? Mustn't these points then be assigned to either blue or orange? How would I go about that @miguelriemoliveira?

**Comment by miguelriemoliveira** [comment]
Hi, just a clarification

> I'm only using:
> number_iterations = 100
> number_points = points.shape[0]
> distance_threshold = 0.08

number_points is not a parameter, it comes from the number of points in the point cloud.

**Comment by miguelriemoliveira** [comment]
> As far as clustering based on point to point distance goes, I remember from our conversation the other day. The example you showed me was somewhat similar with the fences(?) on the street.

Yes, it is similar, but there I computer polygons for each cluster, here we don't need the polygons.

> I'll have to look into it. Unfortunately, I think using python-pcl is gonna be a problem... because I can't seem to be able to install it. Not using pip nor from source...

well, that's a problem because it would surely help, did you try [this](https://github.com/strawlab/python-pcl#common-setting) ?

**Comment by miguelriemoliveira** [comment]
Hi, any idea why this message does not appear on the issue? do you use
email reply?

Will answer on github ...

*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by miguelriemoliveira** [comment]
> However, let's suppose I do discard red and green using a parameter for the minimum number of inliers. What then? Mustn't these points then be assigned to either blue or orange? How would I go about that @miguelriemoliveira?

Well they naturally would, I think. 

1.Imagine first plane is the blue, then you cluster and you keep only the blues on the right as the inliers to the first plane. Thus, you only remove the blue points on the right from the point cloud.
2. Now you find the second plane which will be the orange points plus some of the blues points on the left.

You agree it should work?

**Comment by tiagomfmadeira** [comment]
> number_points is not a parameter, it comes from the number of points in the point cloud.

Yes this is true. I meant I use it as an arg for the function. Which I obviously must update each time I call it, since the points themselves change (each iteration I call it using the outliers of the previous one)

**Comment by tiagomfmadeira** [comment]
> 
> 
> > However, let's suppose I do discard red and green using a parameter for the minimum number of inliers. What then? Mustn't these points then be assigned to either blue or orange? How would I go about that @miguelriemoliveira?
> 
> Well they naturally would, I think.
> 
> 1.Imagine first plane is the blue, then you cluster and you keep only the blues on the right as the inliers to the first plane. Thus, you only remove the blue points on the right from the point cloud.
> 2. Now you find the second plane which will be the orange points plus some of the blues points on the left.
> 
> You agree it should work?

Yes. Realizing this I removed the comment.

**Comment by miguelriemoliveira** [comment]
Ah, that's why I could not find it ...
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*
